src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,149 @@
+            throw new IndexOutOfBoundsException("Index not found " + index);
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> requiaredNode = searchNodeByIndex(index);
+            Node<T> newNode = new Node<>(requiaredNode.prev, value, requiaredNode);
+            newNode.next.prev = newNode;
+            if (index == 0) {
+                head = newNode;
+            } else {
+                newNode.prev.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return searchNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> requiredNode = searchNodeByIndex(index);
+        T oldVal = requiredNode.value;
+        requiredNode.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> requiredNode = searchNodeByIndex(index);
+        T removedNode = requiredNode.value;
+        deleteNode(requiredNode);
+        return removedNode;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> requiredNode = head;
+        for (int i = 0; i < size; i++) {
+            if (object == null ? requiredNode.value == object : requiredNode.value.equals(object)) {
+                deleteNode(requiredNode);
+                return true;
+            }
+            requiredNode = requiredNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index not found " + index);
+        }
+    }
+
+    private Node<T> searchNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> requiredNode;
+        if (index < size / 2) {
+            requiredNode = head;
+            for (int i = 0; i < index; i++) {
+                requiredNode = requiredNode.next;
+            }
+        } else {
+            requiredNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                requiredNode = requiredNode.prev;
+            }
+        }
+        return requiredNode;
+    }
```

Comment:
You can simplify it to one loop.

Reply:
all method `searchNodeByIndex`

Reply:
Yes, but then i must always started of head and i don't wont it